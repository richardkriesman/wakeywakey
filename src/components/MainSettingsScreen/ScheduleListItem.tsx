import React, { ReactNode } from "react";
import { Icon, ListItem } from "react-native-elements";

export interface ScheduleListItemProps {
    enabled: boolean;
    title: string;

    onSwitchToggled: (enabled: boolean) => void;
}

export interface ScheduleListItemState {
    enabled: boolean;
}

export class ScheduleListItem extends React.Component<ScheduleListItemProps, ScheduleListItemState> {

    public constructor(props: ScheduleListItemProps) {
        super(props);
    }

    public componentWillMount(): void {
        this.setState({ enabled: this.props.enabled });
    }

    public onSwitchValueChanged(isEnabledNow: boolean) {
        this.forceEnabled(isEnabledNow);
        this.props.onSwitchToggled(isEnabledNow);
    }

    public forceEnabled(e: boolean) {
        this.setState({ enabled: e });
    }

    public render(): ReactNode {
        return (
            <ListItem
                title={this.props.title}
                rightIcon={<Icon name="arrow-forward" type="ionicons"/>}
                switch={{
                    onValueChange: this.onSwitchValueChanged.bind(this),
                    value: this.state.enabled
                }}
            />
        );
    }

}
