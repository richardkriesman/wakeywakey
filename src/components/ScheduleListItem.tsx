/**
 * @module components
 */

import React, { ReactNode } from "react";
import { ListItem } from "./list/ListItem";

export interface ScheduleListItemProps {
    enabled: boolean;
    title: string;

    onPress: () => void;
    onSwitchToggled: (enabled: boolean) => void;
}

export interface ScheduleListItemState {
    enabled: boolean;
}

export class ScheduleListItem extends React.Component<ScheduleListItemProps, ScheduleListItemState> {

    public constructor(props: ScheduleListItemProps) {
        super(props);
    }

    public componentWillMount(): void {
        this.setState({ enabled: this.props.enabled });
    }

    public onSwitchValueChanged(isEnabledNow: boolean) {
        this.forceEnabled(isEnabledNow).then(() => {
            this.props.onSwitchToggled(isEnabledNow);
        });
    }

    public forceEnabled(e: boolean): Promise<void> {
        return new Promise<void>((resolve) => {
            this.setState({ enabled: e }, () => {
                resolve();
            });
        });
    }

    public render(): ReactNode {
        return (
            <ListItem
                title={this.props.title}
                rightIcon={{ name: "arrow-forward", type: "ionicons" }}
                onPress={this.props.onPress}
                switch={{
                    onValueChange: this.onSwitchValueChanged.bind(this),
                    value: this.state.enabled
                }}
            />
        );
    }

}
