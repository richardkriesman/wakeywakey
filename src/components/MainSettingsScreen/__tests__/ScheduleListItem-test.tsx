import React from "react";
import "react-native";
import renderer, {ReactTestInstance, ReactTestRenderer} from "react-test-renderer";
import {ScheduleListItem} from "../ScheduleListItem";

it("renders title correctly", () => {
    const tree = renderer.create(
        <ScheduleListItem
            enabled={true}
            title="Schedule 420"
            onPress={jest.fn()}
            onSwitchToggled={jest.fn()} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
});

it("sets the toggle switch with props correctly", () => {

    // test with switch disabled
    const disabledTree = renderer.create(
        <ScheduleListItem
            enabled={false}
            title="Schedule 420"
            onPress={jest.fn()}
            onSwitchToggled={jest.fn()} />
    ).toJSON();
    expect(disabledTree).toMatchSnapshot();

    // test with switch enabled
    const enabledTree = renderer.create(
        <ScheduleListItem
            enabled={true}
            title="Schedule 420"
            onPress={jest.fn()}
            onSwitchToggled={jest.fn()} />
    ).toJSON();
    expect(enabledTree).toMatchSnapshot();

});

it("sets the toggle switch with state change correctly", () => {
    const component: ReactTestRenderer = renderer.create(
        <ScheduleListItem
            enabled={false}
            title="Schedule 420"
            onPress={jest.fn()}
            onSwitchToggled={jest.fn()}
        />
    );

    // get instance
    const instance: (ReactTestInstance&ScheduleListItem)|null = component.getInstance() as any;
    expect(instance).toBeDefined();

    // check if switch can be force-enabled
    instance.forceEnabled(true);
    expect(instance.state.enabled).toBe(true); // ideally this should be through a getter
});
