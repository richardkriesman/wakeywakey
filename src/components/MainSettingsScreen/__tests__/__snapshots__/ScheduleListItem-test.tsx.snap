// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders title correctly 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={null}
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "white",
        "borderColor": "#bcbbc1",
        "flexDirection": "row",
        "padding": 14,
        "paddingVertical": 8,
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          Object {
            "backgroundColor": "transparent",
            "fontSize": 17,
          }
        }
        testID="listItemTitle"
        theme={
          Object {
            "colors": Object {
              "disabled": "hsl(208, 8%, 90%)",
              "divider": "#bcbbc1",
              "error": "#ff190c",
              "grey0": "#393e42",
              "grey1": "#43484d",
              "grey2": "#5e6977",
              "grey3": "#86939e",
              "grey4": "#bdc6cf",
              "grey5": "#e1e8ee",
              "greyOutline": "#bbb",
              "platform": Object {
                "android": Object {
                  "error": "#f44336",
                  "primary": "#2196f3",
                  "secondary": "#9C27B0",
                  "success": "#4caf50",
                  "warning": "#ffeb3b",
                },
                "ios": Object {
                  "error": "#ff3b30",
                  "primary": "#007aff",
                  "secondary": "#5856d6",
                  "success": "#4cd964",
                  "warning": "#ffcc00",
                },
              },
              "primary": "#2089dc",
              "searchBg": "#303337",
              "secondary": "#8F0CE8",
              "success": "#52c41a",
              "warning": "#faad14",
            },
          }
        }
      >
        Schedule 420
      </Text>
    </View>
    <View
      width={16}
    />
    <RCTSwitch
      onChange={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "height": 31,
          "width": 51,
        }
      }
      value={true}
    />
    <View
      width={16}
    />
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "justifyContent": "center",
          }
        }
        theme={
          Object {
            "colors": Object {
              "disabled": "hsl(208, 8%, 90%)",
              "divider": "#bcbbc1",
              "error": "#ff190c",
              "grey0": "#393e42",
              "grey1": "#43484d",
              "grey2": "#5e6977",
              "grey3": "#86939e",
              "grey4": "#bdc6cf",
              "grey5": "#e1e8ee",
              "greyOutline": "#bbb",
              "platform": Object {
                "android": Object {
                  "error": "#f44336",
                  "primary": "#2196f3",
                  "secondary": "#9C27B0",
                  "success": "#4caf50",
                  "warning": "#ffeb3b",
                },
                "ios": Object {
                  "error": "#ff3b30",
                  "primary": "#007aff",
                  "secondary": "#5856d6",
                  "success": "#4cd964",
                  "warning": "#ffcc00",
                },
              },
              "primary": "#2089dc",
              "searchBg": "#303337",
              "secondary": "#8F0CE8",
              "success": "#52c41a",
              "warning": "#faad14",
            },
          }
        }
        underlayColor="white"
      >
        <Text />
      </View>
    </View>
  </View>
</View>
`;

exports[`sets the toggle switch with props correctly 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={null}
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "white",
        "borderColor": "#bcbbc1",
        "flexDirection": "row",
        "padding": 14,
        "paddingVertical": 8,
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          Object {
            "backgroundColor": "transparent",
            "fontSize": 17,
          }
        }
        testID="listItemTitle"
        theme={
          Object {
            "colors": Object {
              "disabled": "hsl(208, 8%, 90%)",
              "divider": "#bcbbc1",
              "error": "#ff190c",
              "grey0": "#393e42",
              "grey1": "#43484d",
              "grey2": "#5e6977",
              "grey3": "#86939e",
              "grey4": "#bdc6cf",
              "grey5": "#e1e8ee",
              "greyOutline": "#bbb",
              "platform": Object {
                "android": Object {
                  "error": "#f44336",
                  "primary": "#2196f3",
                  "secondary": "#9C27B0",
                  "success": "#4caf50",
                  "warning": "#ffeb3b",
                },
                "ios": Object {
                  "error": "#ff3b30",
                  "primary": "#007aff",
                  "secondary": "#5856d6",
                  "success": "#4cd964",
                  "warning": "#ffcc00",
                },
              },
              "primary": "#2089dc",
              "searchBg": "#303337",
              "secondary": "#8F0CE8",
              "success": "#52c41a",
              "warning": "#faad14",
            },
          }
        }
      >
        Schedule 420
      </Text>
    </View>
    <View
      width={16}
    />
    <RCTSwitch
      onChange={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "height": 31,
          "width": 51,
        }
      }
      value={false}
    />
    <View
      width={16}
    />
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "justifyContent": "center",
          }
        }
        theme={
          Object {
            "colors": Object {
              "disabled": "hsl(208, 8%, 90%)",
              "divider": "#bcbbc1",
              "error": "#ff190c",
              "grey0": "#393e42",
              "grey1": "#43484d",
              "grey2": "#5e6977",
              "grey3": "#86939e",
              "grey4": "#bdc6cf",
              "grey5": "#e1e8ee",
              "greyOutline": "#bbb",
              "platform": Object {
                "android": Object {
                  "error": "#f44336",
                  "primary": "#2196f3",
                  "secondary": "#9C27B0",
                  "success": "#4caf50",
                  "warning": "#ffeb3b",
                },
                "ios": Object {
                  "error": "#ff3b30",
                  "primary": "#007aff",
                  "secondary": "#5856d6",
                  "success": "#4cd964",
                  "warning": "#ffcc00",
                },
              },
              "primary": "#2089dc",
              "searchBg": "#303337",
              "secondary": "#8F0CE8",
              "success": "#52c41a",
              "warning": "#faad14",
            },
          }
        }
        underlayColor="white"
      >
        <Text />
      </View>
    </View>
  </View>
</View>
`;

exports[`sets the toggle switch with props correctly 2`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={null}
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "white",
        "borderColor": "#bcbbc1",
        "flexDirection": "row",
        "padding": 14,
        "paddingVertical": 8,
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          Object {
            "backgroundColor": "transparent",
            "fontSize": 17,
          }
        }
        testID="listItemTitle"
        theme={
          Object {
            "colors": Object {
              "disabled": "hsl(208, 8%, 90%)",
              "divider": "#bcbbc1",
              "error": "#ff190c",
              "grey0": "#393e42",
              "grey1": "#43484d",
              "grey2": "#5e6977",
              "grey3": "#86939e",
              "grey4": "#bdc6cf",
              "grey5": "#e1e8ee",
              "greyOutline": "#bbb",
              "platform": Object {
                "android": Object {
                  "error": "#f44336",
                  "primary": "#2196f3",
                  "secondary": "#9C27B0",
                  "success": "#4caf50",
                  "warning": "#ffeb3b",
                },
                "ios": Object {
                  "error": "#ff3b30",
                  "primary": "#007aff",
                  "secondary": "#5856d6",
                  "success": "#4cd964",
                  "warning": "#ffcc00",
                },
              },
              "primary": "#2089dc",
              "searchBg": "#303337",
              "secondary": "#8F0CE8",
              "success": "#52c41a",
              "warning": "#faad14",
            },
          }
        }
      >
        Schedule 420
      </Text>
    </View>
    <View
      width={16}
    />
    <RCTSwitch
      onChange={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "height": 31,
          "width": 51,
        }
      }
      value={true}
    />
    <View
      width={16}
    />
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "justifyContent": "center",
          }
        }
        theme={
          Object {
            "colors": Object {
              "disabled": "hsl(208, 8%, 90%)",
              "divider": "#bcbbc1",
              "error": "#ff190c",
              "grey0": "#393e42",
              "grey1": "#43484d",
              "grey2": "#5e6977",
              "grey3": "#86939e",
              "grey4": "#bdc6cf",
              "grey5": "#e1e8ee",
              "greyOutline": "#bbb",
              "platform": Object {
                "android": Object {
                  "error": "#f44336",
                  "primary": "#2196f3",
                  "secondary": "#9C27B0",
                  "success": "#4caf50",
                  "warning": "#ffeb3b",
                },
                "ios": Object {
                  "error": "#ff3b30",
                  "primary": "#007aff",
                  "secondary": "#5856d6",
                  "success": "#4cd964",
                  "warning": "#ffcc00",
                },
              },
              "primary": "#2089dc",
              "searchBg": "#303337",
              "secondary": "#8F0CE8",
              "success": "#52c41a",
              "warning": "#faad14",
            },
          }
        }
        underlayColor="white"
      >
        <Text />
      </View>
    </View>
  </View>
</View>
`;
